// wxflows/dist/lib/generate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateSchema = void 0;
// Copyright IBM Corp. 2023, 2024
const graphql_request_1 = require("graphql-request");
const graphql_1 = require("graphql");
async function generateSchema(url, apikey) {
    const query = (0, graphql_1.getIntrospectionQuery)();
    const requestHeaders = apikey
        ? {
            Authorization: `apikey ${apikey}`,
        }
        : undefined;
    const response = await (0, graphql_request_1.request)({
        url,
        document: query,
        requestHeaders: { 'Content-Type': 'application/json', ...requestHeaders },
    })
        .then(data => {
        return data;
    })
        .catch(error => {
        return error;
    });
    if (!response?.__schema) {
        return {
            error: {
                message: `Error introspecting the endpoint ${url}: \n` +
                    `${JSON.stringify(response?.response?.errors)}`,
            },
        };
    }
    return response;
}
exports.generateSchema = generateSchema;
